test/unit/org/apache/cassandra/net/ConnectionTest.java

```diff
@@ -766,40 +767,84 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+            // max capacity equals to permit-free sendQueueCapcity + the minimun of endpoint and global reserve
+            double maxSendQueueCapacity = outbound.settings().applicationSendQueueCapacityInBytes +
+                                          Double.min(outbound.settings().applicationSendQueueReserveEndpointCapacityInBytes,
+                                                     outbound.settings().applicationSendQueueReserveGlobalCapacityInBytes.limit());
+            int concurrency = 100;
+            int attempts = 10000;
+            int acquireCount = concurrency * attempts;
+            long acquireStep = Math.round(maxSendQueueCapacity * 1.2 / acquireCount / 2); // It is guranteed to acquire (~20%) more
+            // The total overly acquired amount divides the amount acquired in each step. Get the ceil value so not to miss the acquire that just exceeds.
+            long maxFailures = (long) Math.ceil((acquireCount * acquireStep * 2 - maxSendQueueCapacity) / acquireStep); // The result must be in the range of lone
             AtomicLong acquisitionFailures = new AtomicLong();
-            for (int i = 0; i < 100; i++)
-            {
-                executor.submit(() -> {
-                    for (int j = 0; j < 10000; j++)
-                    {
-                        if (!outbound.unsafeAcquireCapacity(acquireStep))
-                            acquisitionFailures.incrementAndGet();
-                    }
+            Runnable acquirer = () -> {
+                for (int j = 0; j < attempts; j++)
+                {
+                    if (!outbound.unsafeAcquireCapacity(acquireStep))
+                        acquisitionFailures.incrementAndGet();
+                }
+            };
+            Runnable releaser = () -> {
+                for (int j = 0; j < attempts; j++)
+                    outbound.unsafeReleaseCapacity(acquireStep);
+            };
 
-                });
-            }
+            // Start N acquirer and releaser to contend for capcaity
+            List<Runnable> invokeOrder = new ArrayList<>();
+            for (int i = 0 ; i < concurrency ; ++i)
+                invokeOrder.add(acquirer);
+            for (int i = 0 ; i < concurrency ; ++i)
+                invokeOrder.add(releaser);
+            // randomize their start order
+            randomize(invokeOrder);
 
-            for (int i = 0; i < 100; i++)
+            try
             {
-                executor.submit(() -> {
-                    for (int j = 0; j < 10000; j++)
-                        outbound.unsafeReleaseCapacity(acquireStep);
-                });
-            }
+                // Reserve enough capacity upfront to ensure the releaser threads cannot release all reserved capacity.
+                // i.e. the pendingBytes is always positive during the test.
+                Assert.assertTrue(outbound.unsafeAcquireCapacity(acquireCount, acquireCount * acquireStep));
```

Comment:
nit: `assertTrue` produces bad error messages; it is mostly better to provide a meaningful message when using `assertTrue`